lane_extraction/src/lane_extraction/window_extractor.cpp

```diff
@@ -0,0 +1,258 @@
+#include <cv_bridge/cv_bridge.h>
+#include <array>
+#include <memory>
+#include <algorithm>
+#include <sensor_msgs/Image.h>
+#include <opencv2/opencv.hpp>
+#include "lane_extraction/utils.h"
+#include "lane_extraction/window_extractor.h"
+
+WindowExtractImpl::WindowExtractImpl()
+    : _imageTransport{_nh}
+{
+    std::string lane_seg_topic;
+    ROS_ASSERT(ros::param::get("~sub_laneseg", lane_seg_topic));
+
+    image_transport::TransportHints transport_hint{"compressed"};
+
+    _laneSegSub = _imageTransport.subscribe(
+        lane_seg_topic,
+        10,
+        [this](const sensor_msgs::ImageConstPtr &msg)
+        {
+            cv_bridge::CvImagePtr cv_ptr;
+            try
+            {
+                cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::MONO8);
+                if (!cv_ptr->image.empty())
+                {
+                    this->_laneSegImage = cv_ptr->image.clone();
+                }
+            }
+            catch (cv_bridge::Exception &e)
+            {
+                ROS_ERROR("Could not convert from '%s' to 'mono8'.", msg->encoding.c_str());
+            }
+        },
+        ros::VoidPtr(), transport_hint);
+}
+
+void WindowExtractImpl::update()
+{
+    if (this->_laneSegImage.empty())
+    {
+        return;
+    }
+
+    cv::Mat birdview, birdTransform;
+    std::tie(birdview, birdTransform) = birdviewTransformation(this->_laneSegImage, birdwidth, birdheight,
+                                                               skyline, offsetLeft, offsetRight);
+
+    birdview(cv::Rect(0, 0, birdview.cols, dropTop)) = cv::Scalar{0};
+
+    left.update(birdview);
+    right.update(birdview);
```

Comment:
> Should we separate preprocessing part to another class? `WindowExtractImpl` should be doing its task only which are `detect` and `tracking`. What if in case we want to use the raw image without bird view transformation. 

Yeah, I think the current implementation assumes that bird-view transformation is used, but we can add configuration, e.g. using the `dynamic-config` package, to use or to not use this transformation.

> If you want `detectImpl` is representative of an algorithm flow ( preprocess -> detect -> track), I suggest moving `birdview` to outside of the detecting and tracking part.

Yes, the sliding window at the moment is a full flow. More specifically, preprocessing with birdview transformation, detection, and tracking. Currently, the preprocessing step is already outside of detect and tracking step as you can see here. As I said above, the birdview step is configurable and I will update it soon.

> Otherwise, please move it to preprocess part to `LaneExtractor` (maybe create a new abstract implement `preprocessImpl`).

I don't think moving preprocessing to a separated class is a good idea at the moment. It might be useful in mixing preprocessing `A` with processing `B`, however, we don't know for sure that the combination is runnable or it makes sense or not.